dnsimple/resource_dnsimple_domain.go

```diff
@@ -0,0 +1,114 @@
+package dnsimple
+
+import (
+	"context"
+	"log"
+	"strconv"
+	"strings"
+
+	"github.com/dnsimple/dnsimple-go/dnsimple"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func resourceDNSimpleDomain() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceDNSimpleDomainCreate,
+		ReadContext:   resourceDNSimpleDomainRead,
+		DeleteContext: resourceDNSimpleDomainDelete,
+
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"account_id": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+			"registrant_id": {
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+			"unicode_name": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"state": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"auto_renew": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+			"private_whois": {
+				Type:     schema.TypeBool,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceDNSimpleDomainCreate(_ context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	provider := meta.(*Client)
+
+	domainAttributes := dnsimple.Domain{
+		Name: data.Get("name").(string),
+	}
+
+	response, err := provider.client.Domains.CreateDomain(context.Background(), provider.config.Account, domainAttributes)
+
+	if err != nil {
+		return diag.Errorf("Failed to create DNSimple Domain: %s", err)
+	}
+
+	data.SetId(strconv.FormatInt(response.Data.ID, 10))
+
+	return resourceDNSimpleDomainRead(nil, data, meta)
+}
+
+func resourceDNSimpleDomainRead(_ context.Context, data *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	provider := meta.(*Client)
+
+	domainID := data.Id()
+
+	response, err := provider.client.Domains.GetDomain(context.Background(), provider.config.Account, domainID)
```

Comment:
This is because how the terraform testing actually works (as I commented in passing [here](https://github.com/dnsimple/terraform-provider-dnsimple/pull/43#issuecomment-890987540) ).

When a test is run, internally the resources method will be called and the testing code is really just a verification of that run that happened in the background. It gave me a major headache as well, because I understood it at first as you do and the tests where failing for "random" reasons I could not understand (until the coin dropped and I saw what's really happening).